rust/src/client.rs

```diff
@@ -0,0 +1,197 @@
+pub enum ClientError {
+    /// Error starting the underlying `Participant`
+    #[error("failed to initialise participant: {0}")]
+    ParticipantInitErr(InitError),
+
+    /// Error from the underlying `Participant`
+    #[error("error arising from participant")]
+    ParticipantErr(PetError),
+
+    /// Error deserialising service data
+    #[error("failed to deserialise service data: {0}")]
+    DeserialiseErr(bincode::Error),
+
+    /// General client errors
+    #[error("unexpected client error")]
+    GeneralErr,
+}
+
+/// A client of the federated learning service
+///
+/// `Client` is responsible for communicating with the service, deserialising
+/// its messages and delegating their processing to the underlying `Participant`.
+pub struct Client {
+    /// Handle to the federated learning `Service`
+    handle: Handle,
+
+    /// The underlying `Participant`
+    participant: Participant,
+
+    /// Interval to poll for service data
+    interval: time::Interval,
+
+    // TODO global model
+
+    id: u32, // NOTE identifier for client for testing; may remove later
+}
+
+impl Client {
+    /// Create a new `Client`
+    ///
+    /// `period`: time period at which to poll for service data, in seconds.
+    /// Returns `Ok(client)` if `Client` `client` initialised successfully
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    pub fn new(period: u64) -> Result<Self, ClientError> {
+        let (handle, _events) = Handle::new();
+        let participant = Participant::new()
+            .map_err(ClientError::ParticipantInitErr)?;
+        Ok(Self {
+            handle,
+            participant,
+            interval: time::interval(Duration::from_secs(period)),
+            id: 0,
+        })
+    }
+
+    /// Create a new `Client` with ID (useful for testing)
+    ///
+    /// `period`: time period at which to poll for service data, in seconds.
+    /// `id`: an ID to assign to the `Client`.
+    /// Returns `Ok(client)` if `Client` `client` initialised successfully
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    pub fn new_with_id(period: u64, handle: Handle, id: u32) -> Result<Self, ClientError> {
+        let participant = Participant::new()
+            .map_err(ClientError::ParticipantInitErr)?;
+        Ok(Self {
+            handle,
+            participant,
+            interval: time::interval(Duration::from_secs(period)),
+            id,
+        })
+    }
+
+    /// Start the `Client` loop
+    ///
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    /// NOTE in the future this may iterate only a fixed number of times, at the
+    /// end of which this returns `Ok(())`
+    pub async fn start(&mut self) -> Result<(), ClientError> {
+        loop {
+            // any error that bubbles up will finish off the client
+            self.per_round().await?;
+        }
+    }
+
+    /// Client duties within a round
+    pub async fn per_round(&mut self) -> Result<Task, ClientError> {
```

Comment:
a minor naming issue: for me this reads like "once per round", but actually this will most likely be awaited multiple times per round, depending on the `interval.tick()`. not sure how to name it better, maybe `during_round()` or sth different like `participate()` or `participate_in_round()`?

Reply:
well, from the perspective of any given `Client` this would indeed be run once per round (assuming normal operation). i'm happy to compromise with `single_round` if you really dislike it though!

Reply:
it's not so much about the intended name but more about what's happening there, to me it looks like `per_round()` will be successfully awaited multiple times per federated learning round in the loop of `start()`:
- a sum participant will be done by the time the sending of the `sum2` message was awaited
- an update participant will be done by the time the sending of the `update` message was awaited
- an unselected participant will be done by the time the round params were awaited

the coordinator will quite likely not have declared an end to the current FL round, therefore `per_round()` will get executed again in the same FL round. maybe i'm missing sth here?

Reply:
ah, see what you mean now. no, your understanding of it is right. previously, the `updater` and `summer` functions would end by calling `unselected`, the idea at the time being it awaits for the global model at the _end_ of the round. in which case, `per_round` does last for the duration of the round. but as the global model is now part of the round parameters (i.e. at the start of the round), recently it was changed to just "return early". since the assumptions have changed now, I think you're right that the name should also change.